pkg/depsdev/v3alpha/api.go

```diff
@@ -172,3 +173,149 @@ func (a *APIv3Alpha) GetPackageVersions(projectName string) (def.PackageVersions
 
 	return response, nil
 }
+
+// GetVersionBatch retrieves a batch of versions for the given version batch requests.
+// The response can be paginated, so the method returns an iterator that allows you to retrieve all the pages content sequentially.
+func (a *APIv3Alpha) GetVersionBatch(req []def.VersionBatchRequest) (*Iterator[def.Version], error) {
+	for _, r := range req {
+		if !input.IsValidPackageManager(r.PackageManager) {
+			return nil, input.ErrInvalidPackageManager
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cIn := getVersionBatch(ctx, a.client, req)
+	iter := Iterator[def.Version]{
+		cIn:     cIn,
+		item:    def.Version{},
+		err:     nil,
+		hasNext: true,
+		cancel:  cancel,
+	}
+
+	return &iter, nil
+}
+
+type versionBatchResponse struct {
+	Responses []struct {
+		Version def.Version `json:"version"`
+	} `json:"responses"`
+	NextPageToken string `json:"nextPageToken"`
+}
+
+func getVersionBatch(ctx context.Context, c *client.Client, req []def.VersionBatchRequest) <-chan batchJob[def.Version] {
+	cJob := make(chan batchJob[def.Version])
+
+	go func() {
+		defer close(cJob)
+
+		requests := []map[string]def.VersionBatchRequest{}
+		for _, r := range req {
+			requests = append(requests, map[string]def.VersionBatchRequest{"versionKey": r})
+		}
+
+		body := map[string]any{
+			"requests":  requests,
+			"pageToken": "",
+		}
+
+		response := versionBatchResponse{
+			NextPageToken: "first",
+		}
+
+		for response.NextPageToken != "" {
+			if err := c.Post(GetVersionBatchPath, body, &response); err != nil {
+				cJob <- batchJob[def.Version]{
+					Err: err,
+				}
+
+				return
+			}
+
+			for _, r := range response.Responses {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					cJob <- batchJob[def.Version]{
+						Item: r.Version,
+					}
+				}
+			}
+
+			body["pageToken"] = response.NextPageToken
+		}
+	}()
+
+	return cJob
+}
+
+// GetProjectBatch retrieves a batch of projects.
+// The response can be paginated, so the method returns an iterator that allows you to retrieve all the pages content sequentially.
+func (a *APIv3Alpha) GetProjectBatch(projectNames []string) (*Iterator[def.Project], error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cIn := getProjectBatch(ctx, a.client, projectNames)
+	iter := Iterator[def.Project]{
+		cIn:     cIn,
+		item:    def.Project{},
+		err:     nil,
+		hasNext: true,
+		cancel:  cancel,
+	}
+
+	return &iter, nil
+}
+
+type projectBatchResponse struct {
+	Responses []struct {
+		Project def.Project `json:"project"`
+	} `json:"responses"`
+	NextPageToken string `json:"nextPageToken"`
+}
+
+func getProjectBatch(ctx context.Context, c *client.Client, projectNames []string) <-chan batchJob[def.Project] {
```

Comment:
the 2 methods`getProjectBatch` and `getVersionBatch` are very similar. I wonder if there is a way to refactor them into a single function...

Reply:
Found a way to extract the duplicate logic. 
If you found this becoming too much magic, I can drop the last commit https://github.com/edoardottt/depsdev/pull/73/commits/f8e0bd0db271c28bba2e170c1de74d5870d36338